### PR TITLE
Return buffers to aravis only after gstreamer no longer references them

### PR DESCRIPTION
**Pull request by TIS-Tim**
Testing showed that the gstreamer pipeline can hold more than one buffer, but arv-viewer currently assumes that the pipeline contains exactly one. It returns the buffer to aravis before gstreamer is done with it, leading to corruption of the displayed image with new data.

This change uses the notification callback from GstBuffer to return the buffer to aravis, which should be the right point in time.

### CODE REVIEW

File: viewer/arvviewer.c [hunk 1]
```diff
@@ -241,8 +241,30 @@ arv_viewer_value_from_log (double value, double min, double max)
...
+	GWeakRef stream;
+	GWeakRef arv_buffer;
+};
+
+static void
+free_arv_buffer_cb (void* param)
+{
+	struct free_arv_buffer_cb_param* cb_param = (struct free_arv_buffer_cb_param*)param;
+
+	ArvStream* stream = g_weak_ref_get (&cb_param->stream);
```
Comment (EmmanuelP) [inline viewer/arvviewer.c]: I did not know this API, thanks !
 Reading the documentation, g_weak_ref converts the weak reference to a normal reference. That probably means there is a reference leak in this function. You should call g_object_unref line 260.

File: viewer/arvviewer.c [hunk 2]
```diff
@@ -241,8 +241,30 @@ arv_viewer_value_from_log (double value, double min, double max)
...
+	ArvBuffer* arv_buffer = g_weak_ref_get (&cb_param->arv_buffer);
```
Comment (EmmanuelP) [inline viewer/arvviewer.c]: arv_buffer should always be valid in this callback, so using g_weak_ref on it is not needed. You should be careful to unref it in the case arv_stream_push_buffer is not called.

File: viewer/arvviewer.c [hunk 3]
```diff
@@ -289,11 +313,19 @@ new_buffer_cb (ArvStream *stream, ArvViewer *viewer)
...
 
-	if (viewer->last_buffer != NULL)
-		arv_stream_push_buffer (stream, viewer->last_buffer);
-	viewer->last_buffer = arv_buffer;
+		gst_app_src_push_buffer (GST_APP_SRC (viewer->appsrc), arv_to_gst_buffer (arv_buffer, cb_param));
+
+		viewer->last_buffer = arv_buffer;
+	}
+	else
+	{
```
Comment (EmmanuelP) [inline viewer/arvviewer.c]: } else {

**Comment by TIS-Tim** [comment]
I updated my branch with your proposed changes. Good catch on the missing unref, that was actually a bad regression.

**Comment by EmmanuelP** [comment]
It looks mostly fine for a merge to master, but I need to perform some tests before.

Thanks Tim.

**Comment by EmmanuelP** [comment]
Sorry for the time it takes, but I still did not take the time to review and test your patches.

**Comment by EmmanuelP** [comment]
Pushed to master with a reference leak fix and debug output. Thanks.

I was not sure about the patches because after a while, the TIS camera became unresponsive. Then I realized its was due to the issue #126 you have already reported.

**Pull request closed without merging**